assembly.py

```diff
@@ -278,45 +284,42 @@ def assemble_trinity(
     else:
         subsamp_bam = util.file.mkstempfname('.subsamp.bam')
+        read_stats = trim_rmdup_subsamp_reads(inBam, clipDb, subsamp_bam, n_reads=n_reads)
 
-    picard_opts = {
-                 'CLIPPING_ATTRIBUTE': tools.picard.SamToFastqTool.illumina_clipping_attribute,
-                 'CLIPPING_ACTION': 'X'
-    }
-
-    read_stats = trim_rmdup_subsamp_reads(inBam, clipDb, subsamp_bam, n_reads=n_reads)
-    subsampfq = list(map(util.file.mkstempfname, ['.subsamp.1.fastq', '.subsamp.2.fastq']))
-    tools.picard.SamToFastqTool().execute(subsamp_bam, subsampfq[0], subsampfq[1], picardOptions=tools.picard.PicardTools.dict_to_picard_opts(picard_opts))
-    try:
-        tools.trinity.TrinityTool().execute(subsampfq[0], subsampfq[1], outFasta, JVMmemory=JVMmemory, threads=threads)
-    except subprocess.CalledProcessError as e:
-        if always_succeed:
-            log.warn("denovo assembly (Trinity) failed to assemble input, emitting empty output instead.")
-            util.file.touch(outFasta)
-        else:
-            raise DenovoAssemblyError(*read_stats)
-    os.unlink(subsampfq[0])
-    os.unlink(subsampfq[1])
-
-    if not outReads:
-        os.unlink(subsamp_bam)
+    with tools.picard.SamToFastqTool().execute_tmp(subsamp_bam, sfx='.subsamp', illuminaClipping=True) as subsampfq:
+        try:
+            tools.trinity.TrinityTool().execute(subsampfq[0], subsampfq[1], 
+                                                outFasta, JVMmemory=JVMmemory, threads=threads)
+        except subprocess.CalledProcessError as e:
+            if always_succeed:
+                log.warn("denovo assembly (Trinity) failed to assemble input, emitting empty output instead.")
+                util.file.touch_empty(outFasta)
+            else:
+                raise DenovoAssemblyError(*read_stats)
 
+    if outReads: shutil.copyfile(subsamp_bam, outReads)
+    if not already_preprocessed_bam: os.unlink(subsamp_bam)
 
 def parser_assemble_trinity(parser=argparse.ArgumentParser()):
     parser.add_argument('inBam', help='Input unaligned reads, BAM format.')
-    parser.add_argument('clipDb', help='Trimmomatic clip DB.')
+    parser.add_argument('clipDb', help='Trimmomatic clip DB.  Ignored if --already_preprocessed_bam.')
     parser.add_argument('outFasta', help='Output assembly.')
     parser.add_argument(
         '--n_reads',
         default=100000,
         type=int,
-        help='Subsample reads to no more than this many pairs. (default %(default)s)'
+        help='Subsample reads to no more than this many pairs. Ignored if --already_preprocessed_bam.  (default %(default)s)'
     )
-    parser.add_argument('--outReads', default=None, help='Save the trimmomatic/prinseq/subsamp reads to a BAM file')
+    parser.add_argument('--outReads', help='Save the trimmomatic/prinseq/subsamp reads to a BAM file')
+    parser.add_argument('--already_preprocessed_bam', default=False, action='store_true',
```

Comment:
Gotcha. `already_realigned_bam` is a file path rather than a boolean though.

We try to be modular and unixy with the viral-ngs python functions, but also expose convenience functions that "pre-chain" functions that are (almost always) called together, like subsampling (or preprocessing) as an inherent part of the `assemble_trinity()`call.

I'd probably vote for keeping each *de novo* assembly preprocessing method as a separate function (exposed via CLI), like how the subsample function is now in `master`, with invocation by default in `assembly(aligner='trinity')` so the most common use case is covered without additional input from the user (with option to disable like you have). This allows savvy users to chain things in a unixy way, but provides "easy" methods for the most common use case.